File: Client/src/dtos/configurations/configurationApplyResultDto.ts
import { ColorConfigurationApplyResultDto } from "./colorConfigurationApplyResultDto";
import { TemperatureConfigurationApplyResultDto } from "./temperatureConfigurationApplyResultDto";

export enum ConfigurationApplyResultDiscriminator {
  TemperatureConfiguration = "temperature",
  ColorConfiguration = "color",
}

export interface ConfigurationApplyResultDto {
  $type: ConfigurationApplyResultDiscriminator;
  devicePath: string;
  succeededToApplyBrightness: boolean;
  applyBrightnessErrors?: string[];
}

export const isTemperatureConfigurationApplyResult = (config?: ConfigurationApplyResultDto): config is TemperatureConfigurationApplyResultDto =>
  {
    return config?.$type == ConfigurationApplyResultDiscriminator.TemperatureConfiguration;
  };

export const isColorConfigurationApplyResult = (config?: ConfigurationApplyResultDto): config is ColorConfigurationApplyResultDto =>
{
  return config?.$type == ConfigurationApplyResultDiscriminator.ColorConfiguration;
};

File: Server/DTOs/Configurations/ConfigurationApplyResultDto.cs
using Swashbuckle.AspNetCore.Annotations;
using System.Text.Json.Serialization;

namespace ScreenTemperature.DTOs.Configurations
{
    [SwaggerDiscriminator("$type")]
    [SwaggerSubType(typeof(TemperatureConfigurationApplyResultDto), DiscriminatorValue = "temperature")]
    [JsonDerivedType(typeof(TemperatureConfigurationApplyResultDto), "temperature")]
    [SwaggerSubType(typeof(ColorConfigurationApplyResultDto), DiscriminatorValue = "color")]
    [JsonDerivedType(typeof(ColorConfigurationApplyResultDto), "color")]
    public abstract class ConfigurationApplyResultDto
    {
        public string DevicePath { get; set; }
        public bool SucceededToApplyBrightness { get; set; }
        public string[] ApplyBrightnessErrors { get; set; }
    }
}
